refactor(env): add zod validation and type-safe env handling

dzx0902 opened the pull request:
## 描述

本 PR 将环境变量访问与校验集中到 `src/config/env.ts`，通过 Zod Schema 提供类型安全与默认值处理，统一 Node/Next 运行时对环境变量的使用边界，并将日志级别、特性开关、数据库连接、Better Auth、Resend/阿里云短信等模块统一接入集中化的配置与辅助函数。

- 严格集中化与归一化：
  - Zod `envSchema` 对空字符串进行预处理，避免构建时的 “too_small” 类错误（如 `BETTER_AUTH_SECRET` 等）`src/config/env.ts:36-94`
  - 轻量特性开关读取，避免客户端/边缘场景交叉字段强校验导致的测试失败：`getFeatureToggles` `src/config/env.ts:190-204`
  - 统一日志级别解析：`getLogLevel`，非法值兜底与警告输出 `src/config/env.ts:247-262`、`src/infra/logging/logger.ts:34-46`
  - CLI/脚本 `.env` 加载器：`ensureNodeEnvLoaded`，为 Drizzle 与脚本显式加载环境 `src/config/load-env.ts:5-9`、`src/db/index.ts:14`
- 数据库配置：
  - `DATABASE_URL` 与 `DATABASE_SSL` 集中读取与推断，结合 `sslmode` 与 `NODE_ENV` 做最终决策 `src/db/index.ts:37-51`
- 运行边界策略：
  - 服务端严格解析（`getParsedEnv`），客户端/边缘使用轻量开关与兜底，避免打包阶段的空值阻塞
- OTP 管道稳定性：
  - 短信与邮箱发送模块读取轻量开关；配置缺失时模拟发送而非抛出校验错误
    - `src/infra/auth/otp-channels.ts:7-15,17-42,44-70`
    - `src/infra/communications/resend.client.tsx:20-33,52-91,99-164`
    - `src/infra/communications/aliyun-sms.client.ts:70-148`
- 日志模块整合与兜底：
  - 当 `LOG_LEVEL` 非法或解析失败时，降级为开发 `debug`/生产 `info` 并输出警告 `src/infra/logging/logger.ts:34-46`
- 框架钩子与中间件接入统一环境判断：
  - `instrumentation.ts:10-18`
  - `src/middleware.ts:73-79`

### 代码定位参考

- 日志级别解析：`src/config/env.ts:247`、`src/infra/logging/logger.ts:34`
- CLI 环境加载：`src/config/load-env.ts:5`
- Drizzle 读取环境：`src/db/index.ts:16`、`drizzle.config.ts:7-15`
- 轻量特性开关：`src/config/env.ts:190-204`
- Turnstile 配置检测：`src/config/env.ts:301-306`
- 安全通知使用集中化环境：`src/infra/notifications/auth-notification.handler.ts:9-13`

## 相关问题

关闭 # 205（如需关联 Issue，请在合并前补充编号）

## 变更类型

- [ ] Bug 修复（不会破坏现有功能的非破坏性变更）
- [ ] 新功能（添加功能的非破坏性变更）
- [ ] 破坏性变更（会导致现有功能无法正常工作的修复或功能）
- [ ] 文档更新
- [ ] 性能改进
- [x] 代码重构
- [ ] CI/CD 改进

## 测试

- 建议本地验证命令（使用 `pnpm`）：
  - `pnpm lint`（Biome 规则与边界检查）
  - `pnpm build`（验证构建通过）
  - `pnpm test:run`（运行 Vitest 全量）
  - `pnpm e2e`（Playwright E2E）
- 注意事项：
  - `.env.test` 需包含有效的 `DATABASE_URL`（含用户名与密码），否则会出现 PG SASL 密码错误
  - 在测试环境，`LOG_LEVEL` 默认解析为 `silent`，可通过显式设置覆盖

- [ ] 我已经在本地测试了这些更改
- [ ] 我已经添加了证明我的修复有效或我的功能工作的测试
- [ ] 新的和现有的单元测试在我的更改下都能通过
- [ ] 我已经检查了代码能够无错误地构建

## 截图（如果适用）

添加截图来帮助解释你的更改。

## 检查清单

- [ ] 我的代码遵循此项目的代码风格
- [ ] 我已经对自己的代码进行了自我审查
- [ ] 我已经对代码进行了注释，特别是在难以理解的地方
- [ ] 我已经对文档进行了相应的更改
- [ ] 我的更改没有产生新的警告
- [ ] 我已经添加了证明我的修复有效或我的功能工作的测试
- [ ] 新的和现有的单元测试在我的更改下都能通过

## 代码审查检查清单（供审查者使用）

- [ ] 代码可读且有良好的文档
- [ ] 更改经过了适当的测试
- [ ] 没有明显的性能问题
- [ ] 已解决安全考虑
- [ ] 破坏性变更已得到适当记录

## 附加说明

- 日志级别默认：`test=silent`、`development=debug`、`production=info`（可通过显式 `LOG_LEVEL` 覆盖）`src/config/env.ts:247-262`
- 重要环境变量的空字符串归一化为未设置（如 `BETTER_AUTH_SECRET`、`CRON_SECRET`、`GITHUB_CLIENT_ID/SECRET`），避免构建期校验错误
- `ENABLE_ALIYUN_SMS`/`ENABLE_RESEND` 支持 `enabled/disabled/true/false/""`；生产默认开启，非生产默认关闭（未显式设置时）
- 数据库 SSL 判定优先级：显式 `DATABASE_SSL` > `DATABASE_URL` 中 `sslmode` > `NODE_ENV`（生产启用）`src/db/index.ts:37-51`

dzx0902 committed:
refactor(env): add zod validation and type-safe env handling

- Use Zod for environment variable validation and type coercion
- Add getParsedEnv helper to centralize env parsing
- Extract database-related configuration into a dedicated helper
- Refine log level handling logic
- Introduce unified feature-flag management
- Update related modules to use the new env handling utilities

dzx0902 committed:
Merge branch 'develop' into refactor/utilize-environment-variables

ukeSJTU commented on app/api/chat/route.ts:
```diff
@@ -7,7 +7,7 @@ export const runtime = "edge";
 const openai = createOpenAICompatible({
   name: "inkeep",
-  apiKey: process.env.INKEEP_API_KEY,
+  apiKey: process.env.INKEEP_API_KEY ?? "",
```
我认为这个不需要添加到 src/config/env.ts，因为 inkeep 是给 fumadocs 用的，我们的文档不计划添加本功能，为了避免给其他开发者增加负担，不应该添加到 env

ukeSJTU commented:
你需要处理一下 gemini 在上面给出的建议：
1. 如果你认为 gemini 的建议是合理的，你应该按照它的建议修改代码，并可以在 "Reply ..." 的位置用类似 "Resolved" 等表示你已经处理该问题；
2. 如果你认为 gemini 的建议有问题，你应该在 "Reply ..." 的位置简单说明它的这个建议有什么问题；
3. 无论你采纳 / 不采纳 Gemini 的建议，你都应该在最后点击 Resolve Conversation 表示你已经处理

dzx0902 commented:
OKOK，昨天晚上太晚了，CI/CD测试跑完就睡觉了

dzx0902 commented on .env.test:
```diff
@@ -5,7 +5,7 @@
 # Format: postgresql://[username]:[password]@[host]:[port]/[database_name]
 # Example: postgresql://user:password@localhost:5432/nomad_db
-DATABASE_URL=postgresql://localhost:5432/nomad_test
+DATABASE_URL=postgresql://postgres:09022004@localhost:5433/nomad
```
这个是本地没有ignore，这个文件确实不应该上传

dzx0902 commented on src/config/env.ts:
```diff
@@ -1,23 +1,310 @@
+  );
+}
+
+export function hasTurnstileConfig(): boolean {
+  const env = getParsedEnv();
+  const siteKey =
+    (env as any).TURNSTILE_SITE_KEY ?? env.NEXT_PUBLIC_TURNSTILE_SITE_KEY;
+  return Boolean(env.TURNSTILE_SECRET_KEY && siteKey);
+}
```
这个是env.ts文件中的定义缺失，我会在fix中补上

dzx0902 commented on src/infra/auth/better-auth.plugin.ts:
```diff
@@ -65,8 +66,8 @@ export const auth = betterAuth({
-      clientId: process.env.GITHUB_CLIENT_ID as string,
-      clientSecret: process.env.GITHUB_CLIENT_SECRET as string,
+      clientId: (getParsedEnv() as any).GITHUB_CLIENT_ID as string,
+      clientSecret: (getParsedEnv() as any).GITHUB_CLIENT_SECRET as string,
```
这是一个合理的建议，我会和之前提出的问题一起在新的commit中修改

dzx0902 commented on src/config/env.ts:
```diff
@@ -1,23 +1,310 @@
+
+export type EnvType = z.infer<typeof envSchema>;
+
+export function getParsedEnv(): EnvType {
+  const parsed = envSchema.parse(process.env as Record<string, unknown>);
+  return parsed as EnvType;
+}
```
这个最好不要缓存，因为像在测试等一些环境中，vitest会用stuenv动态设置环境变量，所以如果使用缓存机制会导致报错。

dzx0902 commented on src/config/env.ts:
```diff
@@ -1,23 +1,310 @@
+  };
+
+  const sms = normalize(env.ENABLE_ALIYUN_SMS as unknown as string | undefined);
+  if (sms) {
+    const issues: z.ZodIssue[] = [];
+    if (!(env as any).ALIBABA_CLOUD_SMS_SIGN_NAME)
+      issues.push({
+        code: z.ZodIssueCode.custom,
+        path: ["ALIYUN_SMS_SIGN_NAME"],
+        message: "Aliyun SMS requires SIGN_NAME when enabled",
+      });
+    if (!(env as any).ALIBABA_CLOUD_SMS_TEMPLATE_CODE)
+      issues.push({
+        code: z.ZodIssueCode.custom,
+        path: ["ALIYUN_SMS_TEMPLATE_CODE"],
+        message: "Aliyun SMS requires TEMPLATE_CODE when enabled",
+      });
+    if (issues.length) throw new z.ZodError(issues);
+  }
+
+  const email = normalize(env.ENABLE_RESEND as unknown as string | undefined);
+  if (email) {
+    const issues: z.ZodIssue[] = [];
+    if (!env.RESEND_API_KEY)
+      issues.push({
+        code: z.ZodIssueCode.custom,
+        path: ["RESEND_API_KEY"],
+        message: "RESEND_API_KEY is required when ENABLE_RESEND=enabled",
+      });
+    if (!(env as any).RESEND_FROM_EMAIL)
+      issues.push({
+        code: z.ZodIssueCode.custom,
+        path: ["RESEND_FROM_EMAIL"],
+        message: "RESEND_FROM_EMAIL is required when ENABLE_RESEND=enabled",
+      });
+    if (issues.length) throw new z.ZodError(issues);
+  }
+
+  return { sms, email };
+}
```
却是因为分模块化修改，不同批次修改之间可能会出现所说的重定义现象，我会在新的commit中修改

dzx0902 commented on src/infra/auth/otp-channels.ts:
```diff
@@ -40,15 +22,19 @@ export async function sendAuthPhoneOtp(
 
-  const success = await sendSmsOtp(phoneNumber, code);
-
-  if (!success) {
+  try {
+    const success = await sendSmsOtp(phoneNumber, code);
+    if (!success) {
+      throw new Error("Failed to send SMS");
+    }
+  } catch {
     throw new Error("Failed to send SMS");
   }
```
嗯，这个问题确实需要考虑，我会在新的commit中进行修改

dzx0902 commented on src/infra/communications/resend.client.tsx:
```diff
@@ -52,15 +54,15 @@ export class ResendEmailClient {
-      const fromEmail =
-        process.env.RESEND_FROM_EMAIL || "onboarding@resend.dev";
+      const env = getParsedEnv();
+      const fromEmail = (env as any).RESEND_FROM_EMAIL;
```
这个实际上会受实际测试时的动态设置影响，所以作为后备预防机制保留

dzx0902 commented on src/infra/notifications/auth-notification.handler.ts:
```diff
@@ -2,11 +2,13 @@ import "server-only";
-const resendFrom = process.env.RESEND_FROM_EMAIL || "onboarding@resend.dev";
+const env = getParsedEnv();
+const resendApiKey = env.RESEND_API_KEY;
+const resendFrom = (env as any).RESEND_FROM_EMAIL || "onboarding@resend.dev";
```
具体原因如上，作为后备和保障机制保留，所以不做修改

ukeSJTU reviewed (commented):
你可以所有的内容都修改完成后，重新 request review~

dzx0902 committed:
refactor(auth): improve env checks and error logging

- Simplify Turnstile config checks to use NEXT_PUBLIC_TURNSTILE_SITE_KEY directly
- Add error logging when OTP sending fails
- Refactor GitHub social login config to enable only when required env vars are set

dzx0902 committed:
Merge branch 'refactor/utilize-environment-variables' of https://github.com/ukeSJTU/nomad into refactor/utilize-environment-variables

ukeSJTU commented:
/gemini review

ukeSJTU commented on .env.test:
`.env.test` 文件是我们项目测试环境 env 参考模版文件，如果需要为自己本地测试环境单独设置环境变量，我建议可以创建一个 `.env.test.local`。总之不应该提交修改到 `.env.test`

dzx0902 replied:
就是我之后的两次commit都已经修改回去了，但push的时候好像没有同步到这个pr里，就很奇怪，我看看是什么情况，因为我giithub网站上的分支里感觉好像也已经改回去了

dzx0902 replied:
我又查了一下，filechange里已经没有这个问题了，应该是已经修好了吧

ukeSJTU commented on playwright.config.ts:
这个 BASE_PATH 参数的作用是什么？

dzx0902 replied:
默认的项目运行时的本地地址，是在Node环境下规定的一个环境变量，一般设置为localhost：3000

dzx0902 commented on .env.test:
```diff
@@ -5,7 +5,7 @@
 # Format: postgresql://[username]:[password]@[host]:[port]/[database_name]
 # Example: postgresql://user:password@localhost:5432/nomad_db
-DATABASE_URL=postgresql://localhost:5432/nomad_test
+DATABASE_URL=postgresql://postgres:09022004@localhost:5433/nomad
```
resolved

dzx0902 commented on app/api/chat/route.ts:
```diff
@@ -7,7 +7,7 @@ export const runtime = "edge";
 const openai = createOpenAICompatible({
   name: "inkeep",
-  apiKey: process.env.INKEEP_API_KEY,
+  apiKey: process.env.INKEEP_API_KEY ?? "",
```
route作为网页端的代码文件，这里的INKEEP_API_KEY是NEXT在Edge运行环境下的环境变量，不适合纳入服务端的集中化环境管理体系中来

dzx0902 commented on src/config/env.ts:
```diff
@@ -1,23 +1,310 @@
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).optional()),
+    ENABLE_RESEND: z
+      .union([
+        z.literal("enabled"),
+        z.literal("disabled"),
+        z.literal("true"),
+        z.literal("false"),
+        z.literal(""),
+      ])
+      .optional(),
+    RESEND_API_KEY: z.preprocess(val => {
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).optional()),
+    RESEND_FROM_EMAIL: z.preprocess(val => {
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).default("onboarding@resend.dev")),
+    NEXT_PUBLIC_TURNSTILE_SITE_KEY: z.preprocess(val => {
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).default("1x00000000000000000000AA")),
+    TURNSTILE_SECRET_KEY: z.preprocess(val => {
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).default("1x0000000000000000000000000000000AA")),
+    CRON_SECRET: z.preprocess(val => {
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).optional()),
+    GITHUB_CLIENT_ID: z.preprocess(val => {
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).optional()),
+    GITHUB_CLIENT_SECRET: z.preprocess(val => {
+      if (typeof val === "string" && val.trim() === "") return undefined;
+      return val;
+    }, z.string().min(1).optional()),
```
对于envschema这样的环境变量的类型或者设置规范定义，不适合引用额外的辅助函数来设置

dzx0902 commented on src/infra/auth/otp-channels.ts:
```diff
@@ -40,15 +22,20 @@ export async function sendAuthPhoneOtp(
-  const success = await sendSmsOtp(phoneNumber, code);
-
-  if (!success) {
+  try {
+    const success = await sendSmsOtp(phoneNumber, code);
+    if (!success) {
+      throw new Error("Failed to send SMS");
+    }
+  } catch (error) {
+    logger.error({ err: error }, "Failed to send SMS OTP");
     throw new Error("Failed to send SMS");
   }
```
resolved

dzx0902 commented on src/config/env.ts:
```diff
@@ -1,23 +1,310 @@
+
+export type EnvType = z.infer<typeof envSchema>;
+
+export function getParsedEnv(): EnvType {
+  const parsed = envSchema.parse(process.env as Record<string, unknown>);
+  return parsed as EnvType;
+}
```
因为测试环境中一个用例vitest会利用stuenv函数多次动态加载环境变量，所以每次使用都要调用而不是缓存，这个修改方案不合适，会产生报错

dzx0902 commented on src/infra/communications/aliyun-sms.client.ts:
```diff
@@ -68,8 +69,9 @@ class AliyunSmsClient {
-      const templateCode = process.env.ALIBABA_CLOUD_SMS_TEMPLATE_CODE;
+      const env = getParsedEnv();
+      const signName = (env as any).ALIBABA_CLOUD_SMS_SIGN_NAME;
+      const templateCode = (env as any).ALIBABA_CLOUD_SMS_TEMPLATE_CODE;
```
resolved

dzx0902 commented on src/infra/communications/resend.client.tsx:
```diff
@@ -52,15 +54,15 @@ export class ResendEmailClient {
-      const fromEmail =
-        process.env.RESEND_FROM_EMAIL || "onboarding@resend.dev";
+      const env = getParsedEnv();
+      const fromEmail = (env as any).RESEND_FROM_EMAIL;
```
上面的问题合并到这里解释，这里只是作为一个引入的保障和冗余机制，防止实际运行中的意外出错，所以留存下来，不做修改

dzx0902 commented on src/infra/http/cron-auth.ts:
```diff
@@ -50,7 +51,8 @@ export function verifyCronSecret(request: Request): boolean {
   // Get the expected secret from environment
-  const expectedSecret = process.env.CRON_SECRET;
+  const env = getParsedEnv();
+  const expectedSecret = (env as any).CRON_SECRET;
```
同上 resolved

dzx0902 commented on src/infra/notifications/auth-notification.handler.ts:
```diff
@@ -2,11 +2,13 @@ import "server-only";
-const resendFrom = process.env.RESEND_FROM_EMAIL || "onboarding@resend.dev";
+const env = getParsedEnv();
+const resendApiKey = env.RESEND_API_KEY;
+const resendFrom = (env as any).RESEND_FROM_EMAIL || "onboarding@resend.dev";
```
同上 resolved

dzx0902 committed:
refactor: remove type assertions on env access

Use the type-safe env helper for environment variable access,
replacing previous `(env as any)` assertions.

ukeSJTU commented on src/infra/notifications/auth-notification.handler.ts:
```diff
@@ -2,11 +2,13 @@ import "server-only";
-const resendFrom = process.env.RESEND_FROM_EMAIL || "onboarding@resend.dev";
+const env = getParsedEnv();
+const resendApiKey = env.RESEND_API_KEY;
+const resendFrom = env.RESEND_FROM_EMAIL || "onboarding@resend.dev";
```
这个是不是在 src/config/env.ts 中已经提供默认值了？你再确认一下～

dzx0902 replied:
是的，这个我fix了

ukeSJTU commented on package.json:
```diff
@@ -12,7 +12,7 @@
     "start": "next start",
     "test": "cross-env NODE_ENV=test vitest",
-    "test:run": "cross-env NODE_ENV=test vitest --run --silent",
+    "test:run": "cross-env NODE_ENV=test vitest --run --silent=true",
```
这个我想知道 `--silent` 和 `--silent=true` 有啥区别吗？没有的话建议保持原本的 `--silent`

dzx0902 replied:
目前我加入了LOG_LEVEL这个环境变量之后，我本地运行--silent会报错，这个我不知道其他环境下是不是好的，我搜索之后的修改结果改为了--silent=true，我这边再看一下

dzx0902 committed:
fix(email): remove default RESEND_FROM_EMAIL value

Remove the default value for RESEND_FROM_EMAIL and rely solely
on the value provided via environment variables.

ukeSJTU commented:
Marking this as completed and merged.
